sshfs.rst

```diff
@@ -120,6 +120,13 @@ Options
    :buflimit: Work around OpenSSH "buffer fillup" bug.
    :createmode: Work around broken servers that produce an error when passing a
                 non-zero mode to create, by always passing a mode of 0.
+   :readdir: Work around file manager used that keeps dir open while
+             user add/remove files/dirs, that produce an error - all dirs
+             become empty for a while or until remount.
+             This happen because handle cached after opendir() but readdir()
+             does not use offset.
+             Workaround converts readdir() into "getdir()": opendir() and
+             releasedir() not exported to fuse; offset set to 0.
```

Comment:
Hmm. Have you thought about storing the last last returned readdir option, and (on the next readdir) check if it corresponds to the stored value? If it matches, readdir can work as before. If it doesn't, it should log a warning and reset the stream. That behavior would make sense in all cases, so it wouldn't even need its own option.

Reply:
Nope, this will not work.
https://github.com/openssh/openssh-portable/blob/4fa9e048c2af26beb7dc2ee9479ff3323e92a7b5/sftp-server.c#L1076
https://www.freebsd.org/cgi/man.cgi?query=readdir&apropos=0&sektion=0&manpath=FreeBSD+12.1-RELEASE+and+Ports&arch=default&format=html

After ```readdir()``` finish you can only ```closedir()```, that is why I enable this workaround for all platforms.
No way to use ```readdir()``` it other way except ```getdir()```.

Reply:
Sorry, I don't follow. I am talking about FUSE's readdir(), not the one in SFTP. Here is the pseudo code:

```
void readdir(..., off_t pos) {
     static int last_offset;
     if (pos != last_offset) {
        log_warning("can't seek in readdir over SFTP, resetting stream to beginning")
        closedir(); opendir();
   }
    // readdir as before
   last_offset = <offset from last returned entry>
}    
```

Does this make more sense?

Reply:
No, look at code in openssh, you will see that after ```readdir()``` handle cant be used any more.
On sshd server side it read all and send all to client, and no way to reset position on dir handle.
```offset``` other than 0 is not allowed in sshfs - there is few asserts for this.

Reply:
What do you mean with "after readir()"? SFTP certainly allows for more than one readdir call (cf. https://tools.ietf.org/html/draft-ietf-secsh-filexfer-13#section-8.2.2). If this doesn't work in the given OpenSSH code, then this sounds like a bug to me (however, withI don't see the problem in the code on a cursory look).

Reply:
Yes, it allow, but only until ```SSH_FX_EOF``` returned, after that dir handle can not be used any more.

Reply:
Sorry, I don't think we are communicating well. I am saying that in the case where you have NOT encountered `SSH_FX_EOF` AND the FUSE client requests the next directory entry, there should be no need to close and re-open the stream.